[REEF-1481] Develop basic metrics APIs for REEF.Net

dkm2110 opened the pull request:
This addressed the issue by
  * adding the APIs for metrics in Org.Apache.REEF.Common.

JIRA:
  [REEF-1481](https://issues.apache.org/jira/browse/REEF-1481)

dkm2110 commented:
@markusweimer  @shravanmn  Can you have a look?

andrey-me commented:
>  	 	 { F 1 A 3 4 9 5 C - 2 C B D - 4 F 3 B - A E D C - 9 C 1 A 4 3 D 9 F 2 3 8 } . R e l e a s e | x 6 4 . B u i l d . 0   =   D e b u g | x 6 4 

looks like temp debug work around. did you want to include this change?

---
Refers to: lang/cs/Org.Apache.REEF.sln:301 in 538eef1. [](commit_id = 538eef13dbe9daecfa0ae159d7df73cfc3e66ac0, deletion_comment = False)

andrey-me commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/AbstractMetric.cs:
```diff
@@ -0,0 +1,131 @@
+        /// </summary>
+        /// <param name="info">Meta-data for the metric</param>
+        /// <param name="value">The value of the metric. It is assumed that user 
+        /// or application has converted float/double values to long via some logic.</param>
```
>or application has converted float/double values to long via some logic [](start = 12, length = 71)

why not to make it generic class? doesn't look like there is any dependency on long type

dkm2110 replied:
So, later on in places like IMetricsRecord and elsewhere where we want to store a collection of different metric types (double, float, int, long)  it's much easier to do that as a single collection, although this can be circumvented by making AbstractMetrics derive from another interface/class with no generic types. Moreover, in case we put it as a generic type, we leave the burden of writing the codecs, string representations and other converters of different types to users. In reality, in all cases we will end up using two kinds of types only - long and double. So I do not want to over complicate this.

andrey-me commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/AbstractMetric.cs:
```diff
@@ -0,0 +1,131 @@
+        /// Type of metric - counter or gauge. Filled in by exact 
+        /// metric type.
+        /// </summary>
+        public abstract MetricType TypeOfMetric { get; }
```
>TypeOfMetric [](start = 35, length = 12)

should the type be part of the metric info?

andrey-me replied:
noticed that metric info is used for tags as well, where type is irrelevant

---
In reply to: [70893489](https://github.com/apache/reef/pull/1073#discussion_r70893489) [](ancestors = 70893489)

andrey-me commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/AbstractMetric.cs:
```diff
@@ -0,0 +1,131 @@
+        /// Accepts a visitor interface
+        /// </summary>
+        /// <param name="visitor">Metrics visitor interface.</param>
+        public abstract void Visit(IMetricsVisitor visitor);
```
>public abstract void Visit(IMetricsVisitor visitor); [](start = 8, length = 52)

if this class is generic and metric visitor generic - then  this doesn't need to be abstract

dkm2110 replied:
This class is still useful to know some very metric specific information - for example, whether it is gauge, counter etc.

andrey-me commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/MetricsTag.cs:
```diff
@@ -0,0 +1,111 @@
+        /// <returns>The string representation of the tag.</returns>
+        public override string ToString()
+        {
+            return "Tag Information: " + _info + ", Tag Value: " + _value;
```
>"Tag Information: " + _info + ", Tag Value: " + _value; [](start = 19, length = 55)

can use string.format instead?

dkm2110 replied:
yes

dkm2110 replied:
yes

andrey-me commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/MetricsTag.cs:
```diff
@@ -0,0 +1,111 @@
+    /// Metrics tag class. Tags can be used to group records together at source, sink or other 
+    /// appropriate places. They can also be used in MetricsFilter to filter records at higher level.
+    /// </summary>
+    public sealed class MetricsTag : IMetricsInfo
```
>MetricsTag [](start = 24, length = 10)

so far looks very similar to AbstractMetric  class except value is string, not long....maybe there is opportunity to reuse some code? create generic an reuse for both cases?

dkm2110 replied:
Ok ... will refactor

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/AbstractMetric.cs:
```diff
@@ -0,0 +1,131 @@
+    /// Abstract immutable metric class. All metrics put in the record by the MetricsSource are 
+    /// kept as <see cref="IEnumerable{T}" />.
+    /// </summary>
+    public abstract class AbstractMetric : IMetricsInfo
```
what's the reasoning for the abstract class? Why not make it an interface with a default implementation?

dkm2110 replied:
It can be ... Although many functions below were common. So I decided to make it abstract.

markusweimer replied:
Abstract classes, especially `public` ones, introduce the [Fragile Base Class Problem](https://en.wikipedia.org/wiki/Fragile_base_class) into the code base. Hence, I generally prefer composition over code inheritance. 

In the case at hand, you can make an interface in place of `AbstractMetric` and then have an implementation of that interface that in turn requires an instance of another interface to be instantiated. That latter interface contains the methods left `abstract` in this class.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/AbstractMetric.cs:
```diff
@@ -0,0 +1,131 @@
+
+using System.Collections.Generic;
+
+namespace Org.Apache.REEF.Common.metrics.Api
```
`metrics` -> `Metrics`

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/AbstractMetric.cs:
```diff
@@ -0,0 +1,131 @@
+        /// <returns>The string representation of the metric.</returns>
+        public override string ToString()
+        {
+            return "Metric Information: " + _info.ToString() + ", Metric Value: " + Value;
```
This does not consider `TypeOfMetric `

dkm2110 replied:
One option is to make it abstract and let user substitute the actual value?

markusweimer replied:
Or use `TypeOfMetric` in the `ToString()`?

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/AbstractMetric.cs:
```diff
@@ -0,0 +1,131 @@
+        /// </summary>
+        /// <param name="obj">Object to compare against.</param>
+        /// <returns>True if both represent the same metric.</returns>
+        public override bool Equals(object obj)
```
This does not consider `TypeOfMetric`. Is that intentional?

dkm2110 replied:
In general name and description are supposed to identify the metric uniquely. I am wondering if we want inter-operability between int-long and float-double in which case its ok to not include type.

markusweimer replied:
Wait, `TypeOfMetric` indicates whether it is an `int` or `long`? I was thinking something like `MB` or `% CPU` here. If it is truly for `int`, `float`, ..., I'd strongly consider using C#'s type system for this instead.

dkm2110 replied:
It indicated whether it is counter or gauge.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/AbstractMetric.cs:
```diff
@@ -0,0 +1,131 @@
+        /// <returns>Hash code.</returns>
+        public override int GetHashCode()
+        {
+            var hashCode = this.ToString().GetHashCode();
```
This does not consider `TypeOfMetric`. Is that intentional?

dkm2110 replied:
Same as above.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/DefaultTypeConverter.cs:
```diff
@@ -0,0 +1,57 @@
+    /// Default implementation of <see cref="ITypeConveter"/>.
+    /// Multiplies float and double by 10^precision and then truncates.
+    /// </summary>
+    public class DefaultTypeConverter : ITypeConveter
```
Typo: `ITypeConveter` -> `ITypeConverter`

markusweimer replied:
Also, why do we need type converters?

dkm2110 replied:
This is more of a convenience class for user to convert double and float to long in case they want to tune precision of truncation. This might be useful for ML algorithms.

markusweimer replied:
So, this is custom casting code? In that case, maybe consider [overloading the casting operator](https://stackoverflow.com/questions/1073480/c-custom-casting-to-a-value-type) instead?

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/IMetricsCollector.cs:
```diff
@@ -0,0 +1,40 @@
+        /// </summary>
+        /// <param name="name">Name of the record.</param>
+        /// <returns>Record builder for the record.</returns>
+        IMetricsRecordBuilder AddRecord(string name);
```
Why call these methods `AddRecord`? Shouldn't it be `NewRecord`?

dkm2110 replied:
So the way I was thinking is that this will add only if record does not exist. Otherwise semantically it might just giver us old existing one.

markusweimer replied:
In that case, maybe call the method `GetOrCreateRecord`? Although, I am still confused by the `Add` prefix. Does it add the record to the collector?

dkm2110 replied:
I have changed the name to CreateRecord and documented the semantics also: "The exact semantics of what to do if another record of same name exists - (create another with same name, return existing one, or throw exception) is left to the implementation."

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/IMetricsFilter.cs:
```diff
@@ -0,0 +1,56 @@
+
+using System.Collections.Generic;
+
+namespace Org.Apache.REEF.Common.metrics.Api
```
`metrics` -> `Metrics`. Probably in all files and the file system :)

dkm2110 replied:
Will fix

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/IMetricsRecord.cs:
```diff
@@ -0,0 +1,58 @@
+    public interface IMetricsRecord
+    {
+        /// <summary>
+        /// Gets the timestamp of the metric.
```
What unit of time?

dkm2110 replied:
I was sort of thinking that implementation will sort of call DateTime.ToLongTimeString() method. Instead I can simply make it return DateTime structure and then leave the whole serializaton or representation issue to the sink where it will have the main use?

markusweimer replied:
Either way, document the unit of time here. [This page](https://blogs.msdn.microsoft.com/brada/2004/03/20/seconds-since-the-unix-epoch-in-c/) has a neat trick to create the seconds-since-the-UNIX-epoch commonly used in the hadoop ecosystem.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/IMetricsRecordBuilder.cs:
```diff
@@ -0,0 +1,117 @@
+        /// <param name="info">Meta data of the metric</param>
+        /// <param name="value">Value of the metric</param>
+        /// <returns></returns>
+        IMetricsRecordBuilder AddCounter(IMetricsInfo info, int value);
```
Do we want to make counters separate? Why not have them implement `AbstractMetric`?

dkm2110 replied:
So we will have two sort of counters - MutableCounters (source facing), and ImmutableCounters. Immutable counters implement AbstractMetrics. This function will be used by mutable counter to add it's value. A typical implementation of IMetricsRecordBuilder will then instantiate ImmutableCounter with this value which will then be passed on to Sink.

markusweimer replied:
Ok, does this method perform the `+=` operation? Or does it make the counter known to the system with its initial value?

dkm2110 replied:
Ohh!! this just allows metrics like MutableCounters to provide its value to record builder so that it can create an immutable counter record out of it for consumption purposes.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/IMetricsRecordBuilder.cs:
```diff
@@ -0,0 +1,117 @@
+        /// <param name="info">Meta data of the metric</param>
+        /// <param name="value">Value of the metric</param>
+        /// <returns></returns>
+        IMetricsRecordBuilder AddCounter(IMetricsInfo info, long value);
```
I'm OK with not having `int` versions of these. 

Is the `value` here the current value or the increment? Also, do we want to allow negative values?

dkm2110 replied:
So just have long and double for simplicity?

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/IMetricsSystem.cs:
```diff
@@ -0,0 +1,77 @@
+    public interface IMetricsSystem
+    {
+        /// <summary>
+        /// (Re)Start the whole Metrics system.
```
What happens when I call `Start` multiple times? Is the second time a restart?

dkm2110 replied:
Yes second time is going to be a restart. I am not sure if somebody calls restart without  calling stop (two successive starts) if we should throw error.

markusweimer replied:
Then please document that.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/ITypeConveter.cs:
```diff
@@ -0,0 +1,42 @@
+    /// <see cref="IMetricsRecordBuilder"/> implementation can use it for 
+    /// converting float and double metric types to long.
+    /// </summary>
+    public interface ITypeConveter
```
typo, should be `ITypeConverter`

dkm2110 replied:
will fix

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/MetricsException.cs:
```diff
@@ -0,0 +1,62 @@
+        /// </summary>
+        /// <param name="message">User defined message.</param>
+        public MetricsException(string message)
+            : base(MessagePrefix() + "\n" + message)
```
I'm no fan of newlines in exception messages, as it makes parsing them really hard.

dkm2110 replied:
will fix.

markusweimer commented:
I've done a pass. A couple of overarching thoughts:

  * The API is pretty `string` heavy. Not sure whether this will come back to haunt us, memory wise.
  * Please consider making most classes `sealed`, such that inheriting from them doesn't become part of our API.

Also, the API seems very much like Java expressed in C#. Which is unsurprising, given where we took inspiration from. A couple of ideas come to mind to make better use of C#:

  * Filters could be expressions from `T` to `bool`
  * `MetricsSource` could be an `IObservable<MetricsType>`
  * `MetricsSink` could be `IObserver<MetricsType>`

dkm2110 commented:
@markusweimer 

> Please consider making most classes sealed
Will do that.

> The API is pretty string heavy.
 I am not sure what you exactly mean - Do you mean that storing name and info for each metric, record etc. is memory intensive. Any suggestion there?

> Filters could be expressions from T to bool
I can make all the functions inside the class as expressions.

> MetricsSource could be an IObservable<MetricsType>
Let me think about this and get back.

dkm2110 commented:
@markusweimer 

> MetricsSource could be an IObservable<MetricsType>
   MetricsSink could be IObserver<MetricsType>

I am not sure if MetricsSource and MetricsSink together qualify for IObservable and IObserver interface. Like in metrics2, we also might want the flow of metrics from source to sink to be asynchronous. For example, source puts metrics in a queue and sink in a separate thread will pull metrics from queue. The maximum length of the queue in turn allows us to make sure that too much memory is not consumed by metrics process and also allows to have the logic of specifying what to do if queue is full (for example sources should hang or just drop metrics.).

On the other hand, MetricsSource and actual metrics like counters and gauges might qualify for IObservable IObserver relationships since when MetricsSource is asked to fill in the metrics records, it can call OnNext of different metrics.

Similarly MetricsSink can be IObserver<MetricsRecord> , with appropriate IObservable defined in the default implementations which in the example above would be the queue implementation in which records are being pushed.

markusweimer commented:
> For example, source puts metrics in a queue and sink in a separate thread will pull metrics from queue.

I don't see why this can't be done in Rx. It is precisely what Rx is for: flowing events through a series of operators, one of which could be a queue.

In general, the API is fairly Java-ish. In C#, we could offer neat things like `_counters["BytesRead"] += 7` or `_bytesReadCounter.OnNext(7)` which would look a lot nicer in client code. Would such an API be layered on top of what we are reviewing here? Or is this neatness an intrinsically bad idea?

dkm2110 commented:
@andrey-me  @markusweimer  I have addressed your comments.

@markusweimer I have introduced another layer of MutableMetrics on top of the APIs that has indexers and operator overloads we discussed. Can you have an overall look and see if it looks ok? If yes then I will introduce relevant tests also for that layer. Alternatively, we can split them in two JIRAs.

Also tried addressing your comments regarding using Rx interfaces.

dkm2110 commented:
@markusweimer  Can you review it?

andrey-me commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/ImmutableMetricImpl.cs:
```diff
@@ -0,0 +1,163 @@
+namespace Org.Apache.REEF.Common.metrics.Api
+{
+    /// <summary>
+    /// Base implementation of <see cref="IImmutableMetric"/>. Takes metric speific 
```
>speific [](start = 76, length = 7)

typo :-)

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/MutableCounter.cs:
```diff
@@ -0,0 +1,114 @@
+        /// </summary>
+        /// <param name="info">Meta-data info. of the metric.</param>
+        /// <param name="initValue">Initial counter value.</param>
+        public MutableCounter(IMetricsInfo info, long initValue)
```
Why is this `public`? Wouldn't an instance of this always be returned by methods in `DefaultMetricsSourceImpl`?

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/MutableCounter.cs:
```diff
@@ -0,0 +1,114 @@
+        /// Counter Constructor. Initializes counter to zero.
+        /// </summary>
+        /// <param name="info">Meta-data info. of the metric.</param>
+        public MutableCounter(IMetricsInfo info) : this(info, 0)
```
Same as above.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/MutableCounter.cs:
```diff
@@ -0,0 +1,114 @@
+        /// Counter Constructor. Initializes counter to zero.
+        /// </summary>
+        /// <param name="name">Name of the counter.</param>
+        public MutableCounter(string name)
```
Same.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/MutableCounter.cs:
```diff
@@ -0,0 +1,114 @@
+        /// <param name="counter">Counter to update.</param>
+        /// <param name="delta">Value by which to update.</param>
+        /// <returns>The updated counter. Reference to input itself.</returns>
+        public static MutableCounter operator +(MutableCounter counter, long delta)
```
This allows `+` operations. Do we support them? Is there a way to only overwrite `+=`?

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/IMutableMetric.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace Org.Apache.REEF.Common.Metrics.MutableMetricsLayer
+{
+    internal interface IMutableMetric : IObserver<SnapshotRequest>
```
Could this also be an `IObservable<MetricsType>`? In other words: What's the design tradeoff between making this push out its updates vs. them being pulled from the outside?

markusweimer replied:
Implementation classes of this interface are `public`, which seems backwards. Wouldn't we expect the interface to be `public`, but the implementations to be `internal`?

dkm2110 replied:
> Could this also be an IObservable<MetricsType>? In other words: What's the design tradeoff between making this push out its updates vs. them being pulled from the outside?

I did not get this comment. Why do we need it to be IObservable<MetricsType> ?  Who is going to be observer?  Also I did not understand what this has to do with push vs. pull based mechanism.

Regarding push vs. pull, in push based mechanism we would need to have timers in each of the MetricSources than having a single timer in IMetricsSystem. IMetricsSystem will then be the observer to which these changes will be pushed in to. So the whole snapshot of the system is decentralized in some sense. Then the question arises when shall we push to sinks - should it be whenever we receive metrics from any source or again it should be governed by some timer in metrics system.

In pull based mechanism, we will have a single centralized control in metrics system. At some fixed time intervals it will pull metrics from registered sources and then push them in to the sinks. 

I would prefer a pull based mechanism if there are no strong arguments in favor of push.

markusweimer replied:
Thanks. You *did* get the question, after all. You answered it :)

So, whenever `OnNext(SnapshotRequest)` gets called, a `IMutableMetric` is supposed to send its current value? What do we gain by that compared to a method `T GetCurrentSnapshot()`, which is called by the class that holds references to `IMutableMetric` instances?

dkm2110 replied:
> So, whenever OnNext(SnapshotRequest) gets called, a IMutableMetric is supposed to send its current value?

Yes, IMutableMetric implementation like Counters will call IMetricRecordBuilder's appropriate function, forexample: AddCounter(info, long value).

> What do we gain by that compared to a method T GetCurrentSnapshot(), which is called by the class that holds references to IMutableMetric instances?

There might be cases while filtering or after creating record where we might want to preserve information like whether it was a counter, gauge or something else. In that case, it would be useful to maintain this information. It is not clear to me how GetCurrentSnapshot() will do it unless the class holding reference maintain the list as counter, gauges etc.,  rather than as interface. Moreover, one more important points, certain metrics like Stats and Rates are actually combination of metrics. In that case return type T will be much more complicated. Hence, I would prefer this sort of function API. 

Does that make sense?

markusweimer replied:
I see. So, the `IMutableMetric` knows how to perform the snapshot, and the event triggers it?

dkm2110 replied:
yes

markusweimer replied:
Ok. What do we gain from having `OnNext(SnapshotRequest)` vs. `DoSnapshot()` or such?

dkm2110 replied:
So we need to pass a recordbuilder so that the metric can be added to the record and other arguments, for example: whether to log the value even if it has not changed. SnapshotRequest captures that....Are you asking why make it an observer instead of simple function call ? We can do either, but I thought it to be more appropriate given timer will  trigger these chains of events from metricsystem to source to metric for getting the snapshot.

What do you think?

markusweimer replied:
Ah, I get it now. Either way is fine with me.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/MutableMetricContainer.cs:
```diff
@@ -0,0 +1,118 @@
+    /// like counters, gauge etc.
+    /// </summary>
+    /// <typeparam name="T">Type of mutable metric.</typeparam>
+    public sealed class MutableMetricContainer<T> where T : MutableMetricBase
```
Why `public`?

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/MutableMetricContainer.cs:
```diff
@@ -0,0 +1,118 @@
+        /// <param name="metricFromName">A function that creates mutable metric from name.</param>
+        /// <param name="metricFromNameDesc">A function that creates mutable metric from nme and desc.</param>
+        /// <param name="provider">Observale to which new metric should subscribe to.</param>
+        public MutableMetricContainer(Func<string, T> metricFromName,
```
Why `public`?

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/MutableMetricContainer.cs:
```diff
@@ -0,0 +1,118 @@
+        /// <param name="name">Name of the metric,</param>
+        /// <param name="desc">Description of the metric,</param>
+        /// <returns>Existing or new metric.</returns>
+        public T this[string name, string desc]
```
What's the relationship of `desc` and `name`? Is `container["myMetric"] == container["myMetric", "MyDescription"] == container["myMetric", "MyOtherDescription"] `? The code suggests that. Why would we have both parameters, then?

dkm2110 replied:
So if metric does not exit, in the beginning when we want to create it for first time we might also want to give description of the metric apart from name. That why I decided to have double indexing. However, name is the unique way by which we identify the metric. Personally, I am not satisfied with it. An alternative might be to ask users to always explicitly create and add metric to the source but subsequent calls can be via index. Another alternative is to assume that via indexing descriptions cannot be provided and we can support additional API like AddCounter(name, descr) etc. 

Any suggestions?

markusweimer replied:
This means that the first time the index operator is called, the one with two parameters is to be used while at all other times, the other one is used? 

If so, I'd prefer a separate method `T CreateMetric(string name, string desc)`. That way, it is obvious when to call which. The description could then become an immutable property of the return value.

dkm2110 replied:
yupp..will do

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/MutableMetricContainer.cs:
```diff
@@ -0,0 +1,118 @@
+                {
+                    if (!_metricsDictionary.ContainsKey(name))
+                    {
+                        throw new MetricsException(string.Format("Metric with name {0} does not exist", name));
```
You can use string interpolation here:

`$"Metric with name {name} does not exist"`

markusweimer commented:
I looked at `MutableMetrics`. A couple of questions:

  * Can you explain what makes them mutable vs. the other metrics? Wouldn't e.g. `MutableCounter` emit immutable records into the metrics system? 
  * I'm a bit concerned with making`MutableCounter` a public type users of the API interact with, as opposed to calling it e.g. `ICounter`. At the very least, we should have that layer of interfaces for the API not to bind user code to implementation classes.
  * There are many `public` classes in this PR. Which makes me nervous, as it means that we commit to those classes in our public APIs. Can you please do a pass and make everything `internal` that you think can be `internal`?
  * [C# doesn't allow the assignment operator to be overloaded](https://msdn.microsoft.com/en-us/library/8edha89s(v=vs.71).aspx). Hence, overloading `+` is dangerous. Also, this currently overloads `+` in a way where the left operand gets mutated, which violates the expectations of such an operator. We might have to drop this nice API simplification.
  * What changed in the solutions file?

dkm2110 commented:
> Can you explain what makes them mutable vs. the other metrics? Wouldn't e.g. MutableCounter emit immutable records into the metrics system?

So, Mutable here means users can modify them. Internally, when IMetricsRecordBuilder functions will be called for example by MutableCounter, it will be stored as IMMutableCounter type of object.

> I'm a bit concerned with makingMutableCounter a public type users of the API interact with, as opposed to calling it e.g. ICounter

I had to do this since operator overloading functions are static and cannot be part of interface. Now that we plan to get rid of them I can do this.

> There are many public classes in this PR. Which makes me nervous, 

Will do a pass.

> C# doesn't allow the assignment operator to be overloaded. Hence, overloading + is dangerous

I propose removing this operator overloading.

> What changed in the solutions file?

Nothing should have. Will take care.

dkm2110 commented:
@markusweimer  replied to your overall main concerns...will send another PR soon.

markusweimer commented:
> So, Mutable here means users can modify them. Internally, when IMetricsRecordBuilder functions will be called for example by MutableCounter, it will be stored as IMMutableCounter type of object.

In that case, I suggest to clean up the user facing names in the API. We could rename `MutableCounter` `ICounter` and the values collected by it (which would now be called simply `Counter`, as they are immutable) to `CounterValue`. Does this make sense to you?

markusweimer commented:
@dkm2110 I've responded to your comments. I agree with the ones I did not respond to.

dkm2110 commented:
@markusweimer 

> In that case, I suggest to clean up the user facing names in the API. 

Agreed.

dkm2110 commented:
@markusweimer @andrey-me   I have pushed in required changes. Please have a look.

markusweimer commented on lang/cs/Org.Apache.REEF.Common.Tests/Org.Apache.REEF.Common.Tests.csproj:
```diff
@@ -55,14 +61,19 @@ under the License.
   <ItemGroup>
     <None Include="packages.config" />
   </ItemGroup>
+  <ItemGroup />
```
This seems unneeded.

markusweimer commented on lang/cs/Org.Apache.REEF.Common.Tests/metrics/DefaultMetricSourceTests.cs:
```diff
@@ -0,0 +1,265 @@
+        }
+
+        /// <summary>
+        /// Tests various functionalities of <see cref="DefaultMetricsSourceImpl"/>
```
Can these be factored into multiple smaller, more focused tests? That way, test failure can be more easily debugged.

dkm2110 replied:
Done ... Also I have retained part of this test as a separate long test to make sure that adding different types of metrics do not lead to some weird bugs.

markusweimer commented on lang/cs/Org.Apache.REEF.Common.Tests/metrics/ImmutableMetricTest.cs:
```diff
@@ -0,0 +1,94 @@
+        /// Tests functions in <see cref="ImmutableMetricsImpl"/>
+        /// </summary>
+        [Fact]
+        public void TestImmutableMetricCommonFunctions()
```
Same: Please refactor into multiple tests that each test a more focused feature.

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common.Tests/metrics/SnapshotRequestTest.cs:
```diff
@@ -0,0 +1,97 @@
+
+namespace Org.Apache.REEF.Common.Tests.Metrics
+{
+    public sealed class SnapshotRequestTest
```
Add documentation

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common.Tests/metrics/SnapshotRequestTest.cs:
```diff
@@ -0,0 +1,97 @@
+            Assert.Equal(request.FullSnapshot, false);
+        }
+
+        private class MetricsRecordBuilderTesImpl : IMetricsRecordBuilder
```
Typo `TesImpl` -> `TestImpl`

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common.Tests/metrics/TestMetricsTag.cs:
```diff
@@ -0,0 +1,76 @@
+
+namespace Org.Apache.REEF.Common.Tests.Metrics
+{
+    public sealed class TestMetricsTag
```
Add documentation.

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Org.Apache.REEF.Common.csproj:
```diff
@@ -222,6 +262,7 @@ under the License.
       <Name>Org.Apache.REEF.Wake</Name>
     </ProjectReference>
   </ItemGroup>
+  <ItemGroup />
```
remove

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/IImmutableMetric.cs:
```diff
@@ -0,0 +1,29 @@
+    /// simply is a combination of functions in <see cref="IImmutableMetricCommonFunctions"/> and 
+    /// <see cref="IImmutableMetricSpecificFunctions"/>.
+    /// </summary>
+    public interface IImmutableMetric : IImmutableMetricCommonFunctions, IImmutableMetricSpecificFunctions
```
Is this a record type or a type that defines object behavior?

dkm2110 replied:
This is the immutable record of the metric taken during snapshot and hence has a timestamp.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/IImmutableMetricCommonFunctions.cs:
```diff
@@ -0,0 +1,44 @@
+    /// Contains functions whose implementation are common to all the metrics. 
+    /// All these functions are implemented once via <see cref="ImmutableMetricsImpl"/>.
+    /// </summary>
+    public interface IImmutableMetricCommonFunctions : IMetricTagCommonFunctions
```
`CommonFunctions` seems like an odd suffix for an interface name. It leaks implementation details into the names.

dkm2110 replied:
@markusweimer  any suggestions for this?

markusweimer replied:
Maybe just `IImutableMetric`?

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/IMetricTagCommonFunctions.cs:
```diff
@@ -0,0 +1,50 @@
+        /// <summary>
+        /// Meta-data of the metric.
+        /// </summary>
+        IMetricsInfo Info { get; }
```
This interface both extends `IMetricsInfo` and has one as a property. What's the rationale for this?

dkm2110 replied:
So I still wanted users to query tag.Name and tag.Desc and hence wanted these functions which are part of IMetricsInfo. On the other hand, I want to create and store IMetricsInfo object once only since I do not want to recreate it if it is asked again and again externally by application or our metrics system.

Alternatively, I can simply remove dependency on IMetricsInfo interface and have it return Name and Description as part of this interface itself.

markusweimer replied:
Inheritance should model a "is a" relationship. So if `IMetricTagCommonFunctions` is a `IMetricsInfo`, you should go with inheritance. If `IMetricTagCommonFunctions` has a `IMetricsInfo` sounds more natural, have the field.

dkm2110 replied:
I will use it as a field.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/IMetricsRecord.cs:
```diff
@@ -0,0 +1,61 @@
+    /// Immutable record of the metric. Represents snapshot of set of metrics
+    /// with a timestamp.
+    /// </summary>
+    public interface IMetricsRecord
```
What's the typical size of one of these records? I'm asking because beyond the metrics, it contains a bunch of strings that can dwarf the payload in size, if we aren't careful.

dkm2110 replied:
I expect the size to be significant. Typically a record if going to be collections of many metrics (loss function, iteration, gradient, norm and other metrics) and not a single one and tags ("ml", "lbfgs" etc.). For example if you look at GetMetrics function in DefaultMetricsSourceImpl it creates a single record by default which means single record per root context and task.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/IMetricsRecordBuilder.cs:
```diff
@@ -0,0 +1,101 @@
+    public interface IMetricsRecordBuilder
+    {
+        /// <summary>
+        /// Adds metrics tag to the record
```
What is a `tag`?

dkm2110 replied:
MetricsTag class...Clarified in the documentation... Are you asking why do we need tags at all....They are useful for end points (like sinks or filters) to do operations. For example an ML application running LR with LBFGS on data might like to tag the record as "ML" and "LBFGS"and "LR".

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/IMetricsSink.cs:
```diff
@@ -0,0 +1,30 @@
+    /// consumes and access the metrics. <see cref="IMetricsSystem"/> will periodically 
+    /// push records from source to sink using the OnNext method of the interface. 
+    /// </summary>
+    public interface IMetricsSink : IDisposable, IObserver<IMetricsRecord>
```
This interface seems redundant. Can't we just use `IObserver`?

dkm2110 replied:
We can ..... I just thought of explicitly calling it out so that people understand that there is a sink. I will remove it.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/MetricsException.cs:
```diff
@@ -0,0 +1,62 @@
+        /// Static message to pre-append to any user message.
+        /// </summary>
+        /// <returns></returns>
+        private static string MessagePrefix()
```
Why not use a `static readonly` field for this?

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/MetricsInfoImpl.cs:
```diff
@@ -0,0 +1,55 @@
+    /// <summary>
+    /// Default implementation of the <see cref="IMetricsInfo"/>
+    /// </summary>
+    public sealed class MetricsInfoImpl : IMetricsInfo
```
Does this class (and its constructor) have to be `public`?

dkm2110 replied:
I expect this class to be used by user very frequently as default IMetricsInfo implementation...Hence providing it to user makes sense.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/MetricsTag.cs:
```diff
@@ -0,0 +1,111 @@
+        /// </summary>
+        /// <param name="info">Meta data for tags.</param>
+        /// <param name="value">Value of the tag.</param>
+        public MetricsTag(IMetricsInfo info, string value)
```
Wouldn't this always be returned by some other method? Hence, it could be `internal`?

dkm2110 replied:
In IMetricsRecordBuilder, we provide function like IMetricsRecordBuilder Add(MetricsTag tag); which means we expect user to create tags. This type of function is useful for instance if user does not want us to create tag internally again and again but would instead create once on his side and reuse it. Let me know if you feel that we should remove this function and make tags internal.

markusweimer replied:
Maybe use a factory then?

dkm2110 replied:
done..added a function for this in IMetricsFactory.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/SnapshotRequest.cs:
```diff
@@ -0,0 +1,56 @@
+        /// </summary>
+        /// <param name="builder">Metrics record builder to be used to add metrics to record.</param>
+        /// <param name="fullSnapshot">If true, record even uncanged metrics.</param>
+        public SnapshotRequest(IMetricsRecordBuilder builder, bool fullSnapshot)
```
Again, why `public`?

dkm2110 replied:
IMutableMetric is public and user might want to extend this interface to write their own metrics class in which case he would need to write what to do when snapshot request comes.

Personally, in my ML algorithms like LBFGS, I might like to make my own UpdateFunction inherit IObserver<SnapshotRequest> interface and directly snapshot iterations, loss function etc. without explicitly creating metrics for them.

markusweimer replied:
I get why the type needs to be `public`. But why the constructor?

dkm2110 replied:
got it done.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/SnapshotRequest.cs:
```diff
@@ -0,0 +1,56 @@
+        /// Constructor. Sets <see cref="FullSnapshot"/> to false.
+        /// </summary>
+        /// <param name="builder">Metrics record builder to be used to add metrics to record.</param>
+        public SnapshotRequest(IMetricsRecordBuilder builder)
```
Why `public`?

dkm2110 replied:
same as above.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/DefaultMetricsFactoryImpl.cs:
```diff
@@ -0,0 +1,206 @@
+        /// <param name="desc">Description of the counter</param>
+        /// <param name="initValue">Initial value of the counter</param>
+        /// <returns>Newly created counter.</returns>
+        public ICounter CreateCounter(string name, string desc, long initValue)
```
You ca merge with the above and use a default value for `initValue`

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/DefaultMetricsFactoryImpl.cs:
```diff
@@ -0,0 +1,206 @@
+        /// <param name="name">Name of the gauge.</param>
+        /// <param name="desc">Description of the gauge.</param>
+        /// <returns>Newly created gauge.</returns>
+        public ILongGauge CreateLongGauge(string name, string desc)
```
Merge with the above, and set a default for `desc`

dkm2110 replied:
The defualt vaue of desc used is name. Hence it cannot be done in this case.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/DefaultMetricsFactoryImpl.cs:
```diff
@@ -0,0 +1,206 @@
+        /// <param name="desc">Description of the gauge</param>
+        /// <param name="initValue">Initial value of the gauge</param>
+        /// <returns>Newly created gauge.</returns>
+        public ILongGauge CreateLongGauge(string name, string desc, long initValue)
```
Again, merge with the above.

dkm2110 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/IMetricContainer.cs:
```diff
@@ -0,0 +1,50 @@
+    {
+        /// <summary>
+        /// Creates metric of type T from name and description.
+        /// Throws exception, if metric with the name already exists.
```
Which exception? Also, this could be documented with the exception tag.

dkm2110 replied:
What do you mean by Exception tag?

dkm2110 replied:
got it will do that.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/metrics/MutableMetricsLayer/IMetricsFactory.cs:
```diff
@@ -0,0 +1,156 @@
+
+namespace Org.Apache.REEF.Common.Metrics.MutableMetricsLayer
+{
+    [DefaultImplementation(typeof(DefaultMetricsFactoryImpl))]
```
Add interface level documentation.

dkm2110 replied:
done

markusweimer commented:
I've done another pass. Some overall comments in addition to the line comments:

  * Please merge multiple method overloads using C#'s default values. That way, the whole interface shrinks considerably.
  * Consider `Nullable` instead of `Optional`
  * This change adds lots of `public` APIs. I left some comments on that, but please do another pass to see what can be made `internal`

dkm2110 commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/IImmutableMetricCommonFunctions.cs:
```diff
@@ -0,0 +1,44 @@
+        /// them as immutable metrics. For a given instance of this interface, either this property 
+        /// or <see cref="NumericValue"/> returns a valid value.
+        /// </summary>
+        Optional<long> LongValue { get; }
```
done

markusweimer replied:
We might also want to take this to the dev list, as have plenty places where we use `Optional`

dkm2110 commented on lang/cs/Org.Apache.REEF.Common/metrics/Api/MetricType.cs:
```diff
@@ -0,0 +1,36 @@
+        /// <summary>
+        /// An arbitrary varying metric, For example loss function, trnasfer rates etc.
+        /// </summary>
+        Gauge
```
Are you saying put comma after Gauge like Gauge,

markusweimer replied:
Yes, sorry. I meant "end with a `,`" :)

dkm2110 commented:
@markusweimer Thanks for the comments. I have addressed many and left comments on some. Once we converge on those I will submit another.

markusweimer commented:
@dkm2110 I've responded to the comments.

dkm2110 commented:
@markusweimer  This is ready for another review.

markusweimer commented:
I'd like to merge this and move on to the implementation. Can you please mark all new `public` APIs as `[Unstable]`? That way we can't accidentally release a version of this and be beholden to it.

dkm2110 commented:
@markusweimer  I have marked public interfaces and classes unstable and also did a rebase. Please test and merge.

markusweimer commented:
testing now
